HW-11

TatyanaV24 opened the pull request:
Задание:
Сделайте следующие компоненты, настраивающиеся в интерфейсе Xcode, которые сразу обновляют свой вид (в сториборде):
a) кнопку, у которой можно изменять ширину и цвет обводки, размер закругления;
b) часы с часовой, минутной и секундной стрелками, у которых можно менять цвет и размер каждой стрелки;
c) собственный аналог UISegmentControl, у которого:
всегда есть два возможных варианта сегмента;
выбранный элемент (по умолчанию — первый) имеет цветную подложку, при нажатии на второй сегмент подложка перемещается под него с анимацией, и наоборот;
в интерфейсе можно задать названия сегментов, цвет обводки и цвет подложки;
есть делегат, через который он будет сообщать о выбранном сегменте.

В целом все условия выполнены, ожидаю комментарии с Вашей стороны.

TatyanaV24 committed:
HW-11

Задание:
Сделайте следующие компоненты, настраивающиеся в интерфейсе Xcode, которые сразу обновляют свой вид (в сториборде):
a) кнопку, у которой можно изменять ширину и цвет обводки, размер закругления;
b) часы с часовой, минутной и секундной стрелками, у которых можно менять цвет и размер каждой стрелки;
c) собственный аналог UISegmentControl, у которого:
всегда есть два возможных варианта сегмента;
выбранный элемент (по умолчанию — первый) имеет цветную подложку, при нажатии на второй сегмент подложка перемещается под него с анимацией, и наоборот;
в интерфейсе можно задать названия сегментов, цвет обводки и цвет подложки;
есть делегат, через который он будет сообщать о выбранном сегменте.

В целом все условия выполнены, ожидаю комментарии с Вашей стороны.

ios-review commented on HW 11/HW 11/ClockView.swift:
```diff
@@ -0,0 +1,107 @@
+    private let roundedView = UIView()
+    
+    override func layoutSubviews() {
+        super.layoutSubviews()
```
1. layoutSubviews используется только для вычисления и изменения frame, bounds, centerPoint или transform свойств для всех подвьюх.
Еще можно изменять свойства, которые зависят от размера самой вью, например, cornerRadius.
layoutSubviews может быть вызвано несколько раз.
Весь остальной код необходимо вынести в другие методы.
Основное правило - свойство должно быть так или иначе связано с размером или положением (координаты).
2. Создание и добавление подвьюх можно осуществлять в инициализаторе https://developer.apple.com/documentation/uikit/uiview/1622488-init
3. Удалить свойство isSetuped.
Свойства жизненного цикла усложняют код.
Рекомендую использовать методы, характеризующие нужный этап жизненного цикла.
В случае первоначальной настройки, таким методом является инициализатор.
Все вью эффективнее создавать в инициализаторе, а потом управлять их состоянием.

ios-review commented on HW 11/HW 11/ClockView.swift:
```diff
@@ -0,0 +1,107 @@
+
+import UIKit
+
+@IBDesignable class ClockView: UIView {
```
Алгоритм вычисления поворота минутной и секундной стрелок работает некорректно.
<img width="1012" alt="Screenshot 2021-02-17 at 18 02 42" src="https://user-images.githubusercontent.com/62285448/108195621-9b2f0280-714a-11eb-8227-3632c83c5c57.png">

ios-review commented on HW 11/HW 11/ViewController.swift:
```diff
@@ -0,0 +1,21 @@
+    }
+
+
+}
```
code-style: пустые файлы не нужны

ios-review commented on HW 11/HW 11/SegmentView.swift:
```diff
@@ -0,0 +1,115 @@
+        sv.topAnchor.constraint(equalTo: self.topAnchor).isActive = true
+        sv.bottomAnchor.constraint(equalTo: self.bottomAnchor).isActive = true
+        sv.leftAnchor.constraint(equalTo: self.leftAnchor).isActive = true
+        sv.rightAnchor.constraint(equalTo: self.rightAnchor).isActive = true
```
Эффективнее и короче:
```
        NSLayoutConstraint.activate([
            sv.topAnchor.constraint(equalTo: self.topAnchor),
            sv.bottomAnchor.constraint(equalTo: self.bottomAnchor),
            sv.leftAnchor.constraint(equalTo: self.leftAnchor),
            sv.rightAnchor.constraint(equalTo: self.rightAnchor)
        ])
```

TatyanaV24 replied:
Почему этот вариант эффективнее? Я встречала подобное описание в решении заданий, т.к. пока ничего подобного на видео уроках не давали, старалась не спешить. Я правильно понимаю, об этом могу прочитать тут - https://developer.apple.com/documentation/uikit/nslayoutconstraint
Приставка NS относится к Objective-C?

ios-review replied:
https://developer.apple.com/documentation/uikit/nslayoutconstraint/1526955-activate
секция discussion

> Typically, using this method is more efficient than activating each constraint individually.

ios-review commented on HW 11/HW 11/SegmentView.swift:
```diff
@@ -0,0 +1,115 @@
+        didSet {
+            updateView()
+        }
+    }
```
Очень дорого при изменении свойства цвета выбранной вью полностью удалять всю иерархию и заново все создавать, конфигурировать.

Эффективнее будет на этапе инициализации создать нужные объекты, а после в нужны методах управлять их состоянием.
```
class SegmentedControl: UIView {
    private let selectorView = UIView(frame: .zero)

     @IBInspectable
     var selectorColor: UIColor = .white {
         didSet {
             selectorView.backgroundColor = selectorColor
         }
     }

     init(frame: CGRect) {
         super.init(frame: frame)
         
         addSubview(selectorView)
     }

     override func layoutSubviews() {
        super.layoutSubviews()
        
        // Обратите внимание, что только в layoutSubviews безопасно использовать frame, так как только в этом методе frame соответствует тому, что на экране.
        // Обязательна к изучению документация метода https://developer.apple.com/documentation/uikit/uiview/1622482-layoutsubviews
        // Особенно Discussion секция.
        let selectorWidth = frame.width / CGFloat(buttonTitles.count)
        selectorView.frame = CGRect(x: 0, y: 0, width: selectorWidth, height: frame.height)
        selectorView.layer.cornerRadius = frame.height / 2
    }
}
```
Подобным образом сделать для всех свойств.

ios-review replied:
При изменении какого свойства все равно придется изменять иерархию вьюх?

ios-review commented on HW 11/HW 11/SegmentView.swift:
```diff
@@ -0,0 +1,115 @@
+    
+    override func draw(_ rect: CGRect) {
+        layer.cornerRadius = frame.height / 2
+    }
```
используйте layoutsubviews

ios-review reviewed (changes requested):
Очень классное решение🔥
Все вью удовлетворяют требованиям.
Молодец👍👍👍

Попробуем сделать код более эффективным?

TatyanaV24 committed:
Обновление задач блок 11

С учетом добавленных комментариев, я внесла правки в код. В итоге Сегмент контроллер нормально не отображается, и я не могу понять почему.
Перечитала информацию по init(frame:), видимо я до конца не понимаю всю прелесть его использования. В данный момент, не понимаю что именно надо исправить, что бы вновь все начало работать корректно.
Буду признательна за пояснения, где именно я наделала ошибок. Спасибо.

ios-review commented on HW 11/HW 11/SegmentView.swift:
```diff
@@ -13,8 +13,13 @@ protocol SegmentViewDelegate {
+    var delegate: SegmentViewDelegate?
+    var selectedIndex: Int = 0
+    var sv = UIStackView()
+    
```
Объекты классов, как и любые другие в жизни, имеют свой жизненный цикл.

Любой объект создается, изменяется под действием различных сил (событий, явлений), а затем погибает.

В swift это выглядит так:
1. Инициализация - создание объекта.
У разных классов могут быть несколько способов создания.
У UIView это -
https://developer.apple.com/documentation/uikit/uiview/1622488-init
https://developer.apple.com/documentation/foundation/nscoding/1416145-init
Первый для создания из кода, второй для создания из сториборда или xib файла.
На этом этапе должно быть создано то, без чего жить нельзя.
А также первоначальная установка.
Например, если бы создавали человека, то нужна голова, тело, руки, ноги.
Для каждой части тела нужно установить цвет, размер, объем первоначальные.

Для UIView - это все ее дочерние вью, которые будут отображаться дальше.
На этом этапе нужно создать эти вью, настроить их, добавить в иерархию.
2. Изменение объекта под действием различных событий.
Для человека это может быть наступление 3 лет. По этому событию человек научается говорить, растет головка, тело, руки, ноги.

Для UIView - это может быть событие отображения на экране. В момент когда вью отображается на экране вызывается метод layoutSubviews для вычисления настоящих размеров и положения всех дочерних вью.
В момент создания вью нет смысла вычислять frame дочерних, так как сама вью еще не отображается и неизвестны размеры этой вью, чтобы посчитать размеры дочерних. А в момент отображения уже можно.
В течение жизни объект изменяется в зависимости от события и тогда, когда это необходимо для решения задачи.
3. Уничтожение объекта - в swift это деинициализация. Сейчас нам это не нужно.

Необходимо научиться видеть весь процесс, состояние объекта в каждый момент, после каждого значимого события, а также, как изменить объект после события, чтобы получить требуемого решение.

Например.
Задача: возможность отображать сегменты вью, задавая их заголовки в сториборде через запятую.
Количество сегментов заранее неопределённо.
Значение по умолчанию - 2 сегмента.

```
class SegmentControl: UIView {
      // Добавляем возможность изменить сегменты через сториборд и свойство.
     @IBInspectable
     var commaSeparatedButtonTitles: String = "1, 2" {
         didSet {
             updateSegments() // После изменения свойства изменяем эти сегменты.
         }
     }
      
     init(frame: CGRect) {
         super.init(frame: frame)
         
          // Добавляем сегменты со значением по умолчанию.
         updateSegments()
     }
     
    func updateSegments() { // В этом методе требуется для каждого заголовка создать дочернюю вью и добавить ее в иерархию.
        // for all title
        // create segment
        // addSubview(segment)
    }

    override func layoutSubviews() {
        // Это особый метод. Его всегда нужно реализовывать отдельно для вычисления frame всех подвью.
       // Только здесь это надо делать.
    }
}
```

ios-review reviewed (changes requested):
Попробуете еще раз разобраться с жизненным циклом объектов?

TatyanaV24 committed:
верные правки

Внесла необходимые изменения в соответствии с комментариями.

ios-review commented on HW 11/HW 11/SegmentView.swift:
```diff
@@ -0,0 +1,122 @@
+        didSet {
+            updateSegments()
+        }
+    }
```
В момент изменения этого свойства кнопки из массива `buttons` уже созданы.
Созданы они на этапе инициализации.
И могут быть еще изменены полностью изменения свойства commaSeparatedButtonTitles.

Кнопки уже созданы к моменту изменения этого свойства.
Это свойство влияет только на цвет текста кнопок в .normal состоянии.
В didSet реализуем только это изменение.
```
    @IBInspectable
    var textColor: UIColor = .lightGray {
        didSet {
            buttons.forEach {
                $0.setTitleColor(textColor, for: .normal)
            }
        }
    }
```

ios-review commented on HW 11/HW 11/SegmentView.swift:
```diff
@@ -0,0 +1,122 @@
+    }
+    
+    @IBInspectable
+    var selectorTextColor: UIColor = .white
```
Это свойство влияет только на цвет текста кнопки в состоянии .selected.
```
    @IBInspectable
    var selectorTextColor: UIColor = .white {
        didSet {
            buttons.forEach {
                $0.setTitleColor(selectorTextColor, for: .selected)
            }
        }
    }
```

ios-review commented on HW 11/HW 11/SegmentView.swift:
```diff
@@ -0,0 +1,122 @@
+            button.addTarget(self, action: #selector(buttonTapped(button:)), for: .touchUpInside)
+            buttons.append(button)
+        }
+        buttons[0].setTitleColor(selectorTextColor, for: .normal)
```
Для легкой смены внешнего вида кнопки в зависимости от ее состояния можно настроить внешний вид для этих состояний, а затем менять само состояние.
Кнопка будет автоматически применять внешний вид для состояния.

Возможные состояния кнопки: https://developer.apple.com/documentation/uikit/uicontrol/state
У функций настройки внешнего вида второй параметр состояние. Например, https://developer.apple.com/documentation/uikit/uibutton/1623993-settitlecolor
Свойство, изменение которого влияет на состояние кнопки: https://developer.apple.com/documentation/uikit/uicontrol/1618203-isselected
Таких свойств несколько 
https://developer.apple.com/documentation/uikit/uicontrol/1618217-isenabled
https://developer.apple.com/documentation/uikit/uicontrol/1618231-ishighlighted

```
        for buttonTitle in buttonTitles {
            let button = UIButton(type: .system)
            button.setTitle(buttonTitle, for: .normal)
            button.setTitleColor(textColor, for: .normal) // Для каждой кнопки устанавливаем цвет текста в нормальном состоянии
            button.setTitleColor(selectorTextColor, for: .selected) // и в выбранном состоянии, так как любая кнопка может быть выбрана
            button.addTarget(self, action: #selector(buttonTapped(button:)), for: .touchUpInside)
            buttons.append(button)
        }
```

ios-review commented on HW 11/HW 11/SegmentView.swift:
```diff
@@ -0,0 +1,122 @@
+                })
+                btn.setTitleColor(selectorTextColor, for: .normal)
+            }
+        }
```
При выборе новой кнопки достаточно для всех кнопок установить isSelected = false, а для той, на которую нажали наоборот true.

Внешний вид для всех кнопок применится автоматически.

ios-review replied:
```
    @objc func buttonTapped(button: UIButton) {
        for (buttonIndex, btn) in buttons.enumerated() {
            if btn == button {
                let selectorStartPosition = frame.width / CGFloat(buttons.count) * CGFloat(buttonIndex)
                selectedIndex = buttonIndex
                delegate?.change(to: selectedIndex)
                UIView.animate(withDuration: 0.3, animations: {
                    self.selector.frame.origin.x = selectorStartPosition
                })
                button.isSelected = true
            } else {
                btn.isSelected = false
            }
        }
    }
```

ios-review commented on HW 11/HW 11/SegmentView.swift:
```diff
@@ -0,0 +1,122 @@
+            sv.leftAnchor.constraint(equalTo: self.leftAnchor),
+            sv.rightAnchor.constraint(equalTo: self.rightAnchor)
+        ])
+    }
```
layoutSubviews используется только для ручного подсчета frame.
Устанавливать констрейнты, то есть autolayout избыточно.
Установить констрейнты можно в момент создания самих вьюх.

И у этой функции надо всегда вызывать super.layoutSubviews().

Потренируемся считать layout?
```
    override func layoutSubviews() {
        super.layoutSubviews()
        
        layer.cornerRadius = self.bounds.height / 2
        
        let selectorWidth = bounds.width / CGFloat( buttonTitles.count)
        selector.frame = CGRect(x: 0, y: 0, width: selectorWidth, height: bounds.height)
        selector.layer.cornerRadius = bounds.height / 2
        
        sv.frame = bounds
    }
```

Для вычисления положения и размеров subviews рекомендуется использовать свойство self.bounds.
Свойство frame может возвращать некорректный результат в случае если свойство transform установлено отличным от .identity.

https://developer.apple.com/documentation/uikit/uiview/1622580-bounds
С bounds такой проблемы нет.

https://developer.apple.com/documentation/uikit/uiview/1622621-frame
```
Warning
If the transform property is not the identity transform, the value of this property is undefined and therefore should be ignored.
```

https://developer.apple.com/documentation/uikit/uiview/1622459-transform

ios-review commented on HW 11/HW 11/ClockView.swift:
```diff
@@ -0,0 +1,104 @@
+        
+        for v in [topMarker, leftMarker, rightMarker, bottomMarker, hourLine, minuteLine, secondLine, roundedView ] {
+            addSubview(v)
+        }
```
layoutSubviews - метод жизненного цикла UIView.
В этом методе можно только считать frame всех дочерних вью.
Еще можно посчитать другие свойства, но значение которых зависит от self.bounds.size.

Добавлять в этом методе новые вью неэффективно и противоречит назначению метода.

Требования к скорости работы layoutSubviews самые строгие.
Требования к производительности самые строгие, когда речь идет про UI.
Такие требования существуют, потому что пользователи любят видеть плавный интерфейс, без подергиваний.
Лирическое отступление:
Если какой-то из методов будет выполняться долго, то может создаться эффект "пропуска кадров", что будет не плавно, с подергиваниями.
На iOS в секунду отображается 60 кадров (экраном телефона).
На один кадр 16,67 миллисекунд.
Если время работы любого пользовательского кода составит больше 16,67 миллисекунд, то кадр будет пропущен.
Пропущен кадр будет, так как процессорное время, необходимое для отрисовки кадра, было занято вычислением другого "тяжелого" кода.

Даже addSubview может в итоге сделать приложение медленным.
Попробуйте проверить на 100, на 1000 вьюх, которые небольшого размера, видны и расположены друг за другом.

Перенести соответствующий код в инициализаторы.
Здесь оставить только тот код, который связан с вычислением frame дочерних вью.
И тот код, который зависит от self.bounds.size.

ios-review reviewed (changes requested):
Работа с жизненным циклом отлично показана на примере SegmentView 🔥

Закрепим для ClockView?
Также добавил несколько упрощений работы с состояниями кнопок.

TatyanaV24 committed:
правки

Внесла правки в соответствии с рекомендациями. Надеюсь все замечания истолковала верно.

TatyanaV24 committed:
.

TatyanaV24 committed:
#9

внесла правки, но код не выводит сообщение в лог.

TatyanaV24 committed:
#10
